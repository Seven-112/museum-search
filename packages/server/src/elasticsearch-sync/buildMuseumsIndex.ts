import { Client } from "elasticsearch";
import { db } from "../sequelize/models";

/**
 * Builds the elasticsearch museums index.
 */
export async function buildMuseumsIndex() {
  const esClient = new Client({
    host: process.env.ES_HOST
  });

  await esClient.ping({});
  console.log("Elasticsearch pinged.");

  // Create museum index if it does not yet exist.
  if (!(await esClient.indices.exists({ index: "museums" }))) {
    await esClient.indices.create({ index: "museums" });
    console.log("museums index created.");
  } else {
    console.log("museums index already exists.");
  }

  await esClient.indices.putMapping({
    body: {
      museum: {
        properties: {
          location: {
            type: "geo_point"
          }
        }
      }
    },
    index: "museums",
    type: "museum"
  });
  console.log("Museum location geo_point mapping created.");

  console.log("Reading museum data from DB.");
  const museums = await db.Museum.findAll().map(data => data.toJSON());
  console.log(`Found ${museums.length} museums in DB`);

  const indexBody: any[] = [];
  museums.forEach(museum => {
    indexBody.push({
      index: { _index: "museums", _type: "museum", _id: museum.id }
    });
    indexBody.push({
      ...museum,
      location: { lat: museum.latitude, lon: museum.longitude }
    });
  });

  if (indexBody.length) {
    console.log("Indexing museum data...");
    await esClient.bulk({
      body: indexBody
    });
    console.log("Museum data indexed.");
  }

  esClient.close();
  db.sequelize.close();
}
