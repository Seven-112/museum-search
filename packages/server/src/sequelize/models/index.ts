import Sequelize from "sequelize";
import { museumFactory } from "./museum";

const env = process.env.NODE_ENV || "development";
// tslint:disable-next-line: no-var-requires
const config = require("../config/config")[env];

const sequelize = new Sequelize(
  config.database,
  config.username,
  config.password,
  config
);

export const db = {
  Museum: museumFactory(sequelize),
  Sequelize,
  sequelize
};

Object.values(db).forEach((model: any) => {
  if (model.associate) {
    model.associate(db);
  }
});
