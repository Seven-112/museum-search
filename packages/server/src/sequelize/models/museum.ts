import { Sequelize, STRING, DOUBLE, Instance } from "sequelize";

export interface MuseumAttributes {
  id?: number;
  name: string;
  legalName?: string;
  alternateName?: string;
  museumType?: string;
  institutionName?: string;
  streetAddress?: string;
  city?: string;
  state?: string;
  zipCode?: string;
  phoneNumber?: string;
  latitude?: number;
  longitude?: number;
}

export type MuseumInstance = Instance<MuseumAttributes> & MuseumAttributes;

export const museumFactory = (sequelize: Sequelize) => {
  const Museum = sequelize.define<MuseumInstance, MuseumAttributes>(
    "Museum",
    {
      name: STRING,
      legalName: STRING,
      alternateName: STRING,
      museumType: STRING,
      institutionName: STRING,
      streetAddress: STRING,
      city: STRING,
      state: STRING,
      zipCode: STRING,
      phoneNumber: STRING,
      latitude: DOUBLE,
      longitude: DOUBLE
    },
    {}
  );
  Museum.associate = models => {
    // associations can be defined here
  };
  return Museum;
};
