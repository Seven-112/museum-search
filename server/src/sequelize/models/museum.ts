"use strict";

import { Sequelize, DataTypes } from "sequelize";

export default (sequelize: Sequelize, DataTypes: DataTypes) => {
  const Museum = sequelize.define(
    "Museum",
    {
      name: DataTypes.STRING,
      legalName: DataTypes.STRING,
      alternateName: DataTypes.STRING,
      museumType: DataTypes.STRING,
      institutionName: DataTypes.STRING,
      streetAddressAdministrative: DataTypes.STRING,
      cityAdministrative: DataTypes.STRING,
      stateAdministrative: DataTypes.STRING,
      zipCodeAdministrative: DataTypes.STRING,
      streetAddressPhysical: DataTypes.STRING,
      cityPhysical: DataTypes.STRING,
      statePhysical: DataTypes.STRING,
      zipCodePhysical: DataTypes.STRING,
      phoneNumber: DataTypes.STRING,
      latitude: DataTypes.INTEGER,
      longitude: DataTypes.INTEGER
    },
    {}
  );
  Museum.associate = models => {
    // associations can be defined here
  };
  return Museum;
};
