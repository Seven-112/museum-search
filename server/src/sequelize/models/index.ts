"use strict";

import Sequelize from "sequelize";
import { museumFactory } from "./museum";

const env = process.env.NODE_ENV || "development";
const config = require(__dirname + "/../config/config.json")[env];

const sequelize = new Sequelize(
  config.database,
  config.username,
  config.password,
  config
);

export const db = {
  sequelize,
  Sequelize,
  Museum: museumFactory(sequelize)
};

Object.values(db).forEach((model: any) => {
  if (model.associate) {
    model.associate(db);
  }
});
