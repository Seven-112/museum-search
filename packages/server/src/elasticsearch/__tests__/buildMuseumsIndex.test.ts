import { db } from "../../sequelize/models";
import { MuseumAttributes } from "../../sequelize/models/museum";
import { buildMuseumsIndex } from "../buildMuseumsIndex";

// Create mock elasticsearch client functions.
const ping = jest.fn();
const exists = jest.fn(async () => false);
const create = jest.fn(async () => {});
const putMapping = jest.fn(async () => {});

const bulk = jest.fn(async () => {});

// Mock the elasticsearch index builder's usage of the elasticsearch Client.
jest.mock("elasticsearch", () => ({
  Client: class {
    ping = ping;

    indices = {
      exists,
      create,
      putMapping
    };

    bulk = bulk;
  }
}));

let findAllSpy: jest.SpyInstance;

beforeAll(() => {
  // Return mock results for Museum's findAll query.
  findAllSpy = jest.spyOn(db.Museum, "findAll").mockImplementation(() => {
    // Create mock museum data.
    const mockData: MuseumAttributes[] = [
      {
        id: 1,
        name: "test museum 1",
        latitude: 10,
        longitude: 3
      },
      {
        id: 2,
        name: "test museum 2"
      }
    ];

    return mockData.map(data => db.Museum.build(data));
  });
});

afterAll(() => {
  // Restore the findAll spy.
  findAllSpy.mockRestore();
});

describe("build-elasticsearch-index", () => {
  it("builds the elasticsearch index", async () => {
    // Execute the index builder.
    await buildMuseumsIndex();

    expect(ping.mock.calls.length).toEqual(1);
    expect(ping.mock.calls[0][0]).toEqual({});

    expect(exists.mock.calls.length).toEqual(1);
    expect(exists.mock.calls[0][0]).toEqual({ index: "museums" });

    expect(create.mock.calls.length).toEqual(1);
    expect(create.mock.calls[0][0]).toEqual({ index: "museums" });

    expect(putMapping.mock.calls.length).toEqual(1);
    expect(putMapping.mock.calls[0][0]).toEqual({
      index: "museums",
      type: "museum",
      body: {
        museum: {
          properties: {
            location: {
              type: "geo_point"
            }
          }
        }
      }
    });

    expect(bulk.mock.calls.length).toEqual(1);
    expect(bulk.mock.calls[0][0]).toMatchSnapshot("bulk operation");
  });
});
