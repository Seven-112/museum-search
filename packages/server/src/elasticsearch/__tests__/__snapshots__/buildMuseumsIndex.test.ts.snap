// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build-elasticsearch-index builds the elasticsearch index: bulk operation 1`] = `
Object {
  "body": Array [
    Object {
      "index": Object {
        "_id": 1,
        "_index": "museums",
        "_type": "museum",
      },
    },
    Object {
      "id": 1,
      "latitude": 10,
      "location": Object {
        "lat": 10,
        "lon": 3,
      },
      "longitude": 3,
      "name": "test museum 1",
    },
    Object {
      "index": Object {
        "_id": 2,
        "_index": "museums",
        "_type": "museum",
      },
    },
    Object {
      "id": 2,
      "location": Object {
        "lat": undefined,
        "lon": undefined,
      },
      "name": "test museum 2",
    },
  ],
}
`;
